crates/ruma-state-res/tests/it/main.rs

```diff
@@ -0,0 +1,249 @@
+            .room_version
+    };
+
+    let mut state_sets = Vec::new();
+    for pdu in &pdus {
+        let mut state_map = StateMap::new();
+        state_map.insert(
+            (
+                pdu.event_type().to_string().into(),
+                pdu.state_key().ok_or("all PDUs should be state events")?.to_owned(),
+            ),
+            pdu.event_id().clone(),
+        );
+        state_sets.push(state_map);
+    }
+
+    let pdus_by_id =
+        pdus.into_iter().map(|pdu| (pdu.event_id().to_owned(), pdu)).collect::<HashMap<_, _>>();
+
+    let mut auth_chain_sets = Vec::new();
+    for pdu in pdus_by_id.values() {
+        auth_chain_sets.push(auth_events_dfs(&pdus_by_id, pdu)?);
+    }
+
+    let fetch_event = |x: &EventId| pdus_by_id.get(x).cloned();
+
+    let resolved_state = resolve(&room_version, &state_sets, auth_chain_sets, fetch_event)?;
+
+    // Reformat this a bit to make the snapshots easier to compare
+    let resolved_state = resolved_state
+        .into_iter()
+        .map(|((kind, state_key), event_id)| {
+            Ok(ResolvedState {
+                kind,
+                state_key,
+                content: to_json_value(pdus_by_id[&event_id].content())?,
+                event_id,
+            })
+        })
+        .collect::<Result<BTreeSet<_>, serde_json::Error>>()?;
+
+    Ok(TestResults { resolved_state })
+}
+
+/// Depth-first search for `auth_events`
+///
+/// # Errors
+///
+/// Fails if `pdus` does not contain a PDU that appears in the recursive `auth_events` of `pdu`.
+fn auth_events_dfs(
+    pdus: &HashMap<OwnedEventId, Pdu>,
+    pdu: &Pdu,
+) -> Result<HashSet<OwnedEventId>, Box<dyn Error>> {
+    let mut out = HashSet::new();
+    let mut stack = pdu.auth_events().cloned().collect::<Vec<_>>();
+
+    while let Some(event_id) = stack.pop() {
+        if out.contains(&event_id) {
+            continue;
+        }
+
+        out.insert(event_id.clone());
+
+        stack.extend(
+            pdus.get(&event_id)
+                .ok_or_else(|| format!("missing required PDU: {}", event_id))?
+                .auth_events()
+                .cloned(),
+        );
+    }
+
+    Ok(out)
+}
+
+/// Create a new snapshot test
+///
+/// # Arguments
+///
+/// * the test function's name
+/// * the JSON file to load PDUs to resolve from
+macro_rules! make_snapshot_test {
+    ($name:ident, $path:expr $(,)?) => {
+        #[test]
+        fn $name() -> Result<(), Box<dyn std::error::Error>> {
+            let crate::TestResults {
+                resolved_state,
+            } = crate::run($path)?;
+
+            insta::with_settings!({
+                description => "Resolved state",
+                omit_expression => true,
+                snapshot_suffix => "resolved_state",
+            }, {
+                insta::assert_json_snapshot!(&resolved_state);
+            });
+
+            Ok(())
+        }
+    };
+}
+
+mod tests;
```

Comment:
This is actually required because of how (declarative) macros work, unfortunately. Modules using the macro have to be defined lexically after the macro.

Reply:
We could also move the macro at the top of the file where it is used :slightly_smiling_face:.

Reply:
I kinda wanted to have `snapshot_tests.rs` contain only definitions and no actual code

Reply:
I've moved the macro and the `mod` statement to the top of `tests/it/resolve.rs`, do you think that's sufficient?